server/src/main/java/org/elasticsearch/search/lookup/SourceLookup.java

```diff
@@ -101,6 +106,15 @@ public void setSegmentAndDocument(LeafReaderContext context, int docId) {
             return;
         }
         this.reader = context.reader();
+        // Lucene stored fields are really optimized for random access and don't
+        // optimize for sequential access - except for merging. So we do a
+        // little hack here and pretend we're going to do merges in order to
+        // get better sequential access.
+        try {
+            this.storedFieldsReader = SlowCodecReaderWrapper.wrap(reader).getFieldsReader().getMergeInstance();
```

Comment:
It looks like we do our best to always have a `CodecReader` here now so maybe we should make that the return type? Or assert that it is one and then hard cast? It'd be a shame to fall back to the "slow" implementation of this a run time accidentally.

Reply:
Getting there requires more work. DLS and FLS currently use reader wrappers that would hide the codec reader.

Reply:
Could you leave a comment then about how we're not using the "slow part" of `SlowCodecReaderWrapper`. I checked the code path a little more closely and I feel much better it. The name is just scary. Really this just wraps the reader so you can call `getMergeInstance` which still returns `this`.